### PR TITLE
Create sample project for Atrium + MPP kotlin.test #81

### PR DESCRIPTION
**Pull request by bsemexan**
----
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle [hunk 1]
```diff
@@ -1,4 +1,4 @@
-description = 'A fluent assertion function API in en_GB with a focus on code completion for the JS platform.'
+description = 'A fluent assertion function API in en_GB with a focus on code completion as js module.'
```
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: fine with me as well but could you change the other places then too please? I prefer to be consistent. You could search & replace with regex enabled:
search: `for the (.*) platform` 
replace with: `as $1 module`
It's enough to search in *.gradle files

File: samples/multiplatform/build.gradle [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        browser {
+        }
+        nodejs {
+        }
+    }
+    // For ARM, should be changed to iosArm32 or iosArm64
+    // For Linux, should be changed to e.g. linuxX64
+    // For MacOS, should be changed to e.g. macosX64
+    // For Windows, should be changed to e.g. mingwX64
+    mingwX64("windows")
```
Comment (robstoll) [inline samples/multiplatform/build.gradle]: That's for the native target, Atrium does not yet support it and thus you can remove it. However, I would like to see an Android target. See
https://kotlinlang.org/docs/reference/building-mpp-with-gradle.html#android-support

File: samples/multiplatform/build.gradle [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    sourceSets {
+
```
Comment (robstoll) [inline samples/multiplatform/build.gradle]: please also add the dependency for commonTest (and androidTest once you have setup the android target). Something along the line of
```
commonTest {
    dependencies {
        implementation kotlin('test-common')
        implementation kotlin('test-annotations-common')
        implementation kotlin('ch.tutteli.atrium:atrium-fluent-en_GB-common:0.9.0-alpha2')
    }
}
```

File: samples/multiplatform/build.gradle [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        jvmTest {
+            dependencies {
+                //implementation kotlin('test')
+                //implementation kotlin('test-junit')
+                implementation("org.junit.jupiter:junit-jupiter-api:5.3.1")
+                implementation("ch.tutteli.atrium:atrium-fluent-en_GB-jvm:0.9.0-alpha2")
```
Comment (robstoll) [inline samples/multiplatform/build.gradle]: jvm modules don't have the -jvm suffix
```suggestion
                implementation("ch.tutteli.atrium:atrium-fluent-en_GB:0.9.0-alpha2")
```

File: samples/multiplatform/src/jsTest/mocha/src/test/kotlin/ch/tutteli/SmokeTest.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
+package ch.tutteli
+
+import ch.tutteli.atrium.api.fluent.en_GB.isLessThan
```
Comment (robstoll) [inline samples/multiplatform/src/jsTest/mocha/src/test/kotlin/ch/tutteli/SmokeTest.kt]: It would have been enough for me if you add one JS runner but good as well.
Feel free to remove one in case you get problems (once you setup the runners to actually execute the tests)

File: samples/multiplatform/src/jvmTest/spek/src/test/kotlin/SmokeTest.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package org.spekframework.spek2
+
+import ch.tutteli.atrium.api.fluent.en_GB.messageContains
+import ch.tutteli.atrium.api.fluent.en_GB.toBe
+import ch.tutteli.atrium.api.fluent.en_GB.toThrow
+import ch.tutteli.atrium.api.verbs.expect
+import org.spekframework.spek2.style.specification.describe
+import java.lang.IllegalArgumentException
+
+object CalculatorTestBySpek : Spek({
```
Comment (robstoll) [inline samples/multiplatform/src/jvmTest/spek/src/test/kotlin/SmokeTest.kt]: same here, it would be enough for me if you only use junit5 as test runner for JVM.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle (hunk 1 above)
Comment (bsemexan) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: i added as a js test was failing with a no description error message

File: samples/multiplatform/src/jvmTest/spek/src/test/kotlin/SmokeTest.kt (hunk 6 above)
Comment (bsemexan) [reply samples/multiplatform/src/jvmTest/spek/src/test/kotlin/SmokeTest.kt]: for this test and mocha test above, I just copied from the ones already accepted, but I will see what I can do.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle (hunk 1 above)
Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: not sure what you mean, this description is used by gradle, I don't see why it should interfere with a js test. Can you elaborate please

Comment (bsemexan) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: I do not recall exactly but it was when I was trying to run a grade build in the multiplatform directory and it was saying 'description can't be null' or something along those lines. When I added it, it stopped throwing that error

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: OK, in this case revert your change please. This modifies an existing description and this cannot be the cause.

Comment (bsemexan) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: reverted change and
`You need to define project.description for publishing (empty or blank is considered to be undefined)
`

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: I guess you deleted it. Just revert to the previous description

Comment (bsemexan) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: it was empty

Comment (bsemexan) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle]: i'll go ahead and copy this
https://github.com/robstoll/atrium/blob/master/apis/fluent-en_GB/atrium-api-fluent-en_GB-js/build.gradle
not sure why when i checked out master it came empty for me

File: samples/multiplatform/build.gradle (hunk 2 above)
Comment (bsemexan) [reply samples/multiplatform/build.gradle]: the android target is a bit mere complicated so might be best to create an issue for it

Comment (robstoll) [reply samples/multiplatform/build.gradle]: fine with me

File: samples/multiplatform/src/jsTest/jasmine/src/test/kotlin/ch/tutteli/SmokeTest.kt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import ch.tutteli.atrium.domain.builders.AssertImpl
+import ch.tutteli.atrium.reporting.RawString
+import ch.tutteli.atrium.reporting.translating.StringBasedTranslatable
+import ch.tutteli.atrium.translations.DescriptionBasic
+import kotlin.test.Test
+
+class SmokeTest {
+    @Test
+    fun toBe_canBeUsed() {
+        assertThat(1).isLessThan(2)
```
Comment (robstoll) [inline samples/multiplatform/src/jsTest/jasmine/src/test/kotlin/ch/tutteli/SmokeTest.kt]: change to 
```suggestion
        assertThat(1).toBe(2) 
```
so that you see the test fails in the build

File: samples/multiplatform/src/jsTest/mocha/src/test/kotlin/ch/tutteli/SmokeTest.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import ch.tutteli.atrium.domain.builders.AssertImpl
+import ch.tutteli.atrium.reporting.RawString
+import ch.tutteli.atrium.reporting.translating.StringBasedTranslatable
+import ch.tutteli.atrium.translations.DescriptionBasic
+import kotlin.test.Test
+
+class SmokeTest {
+    @Test
+    fun toBe_canBeUsed() {
+        assertThat(1).isLessThan(2)
```
Comment (robstoll) [inline samples/multiplatform/src/jsTest/mocha/src/test/kotlin/ch/tutteli/SmokeTest.kt]: change to 
```suggestion
        assertThat(1).toBe(2) 
```
so that you see the test fails in the build

File: samples/multiplatform/src/jvmTest/junit5/src/test/kotlin/ch/tutteli/SmokeTest.kt [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import ch.tutteli.atrium.creating.Expect
+import ch.tutteli.atrium.domain.builders.AssertImpl
+import ch.tutteli.atrium.reporting.RawString
+import ch.tutteli.atrium.reporting.translating.StringBasedTranslatable
+import ch.tutteli.atrium.translations.DescriptionBasic
+
+class SmokeTest {
+    @Test
+    fun toBe_canBeUsed() {
+        assertThat(1).isLessThan(2)
```
Comment (robstoll) [inline samples/multiplatform/src/jvmTest/junit5/src/test/kotlin/ch/tutteli/SmokeTest.kt]: change to 
```suggestion
        assertThat(1).toBe(2) 
```
so that you see the test fails in the build

File: samples/multiplatform/src/jvmTest/spek/src/test/kotlin/SmokeTest.kt [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    describe("Calculator") {
+        val calculator by memoized { Calculator() }
+
+        it("1 + 2 = 3") {
+            val sum = 3
```
Comment (robstoll) [inline samples/multiplatform/src/jvmTest/spek/src/test/kotlin/SmokeTest.kt]: change to 
```suggestion
            val sum = 2
```
so that you see the test fails in the build

**Comment by bsemexan** [comment]
> The tests you have added are still not executed in CI as far as I can see. Make the suggested changes and run `gradlew build` locally to be sure that they are carried out.
> Moreover, you are still missing the test in commonTest (you can copy for instance the SmokeTest of atrium-fluent-en_GB-common)

https://github.com/bugsnag/kotlin-multiplatform-example

From this example, it seems your samples directory would be the 'multiplatform' directory for these to work

**Comment by robstoll** [comment]
The given example project (kotlin-multiplatform-example) has multiple projects where one of them is a multi-platform project (common) and the others are platform specific projects using common. What we want to show here is different. Atrium is our multi-platform project and we want to use it in another multi-platform project. The examples you have made for junit5 and jasmine are the analogous to the platform specific projects in kotlin-multiplatform-example. Do you see the difference?

I suggest you create an example project in Intellij via File -> New... -> Project... -> Kotlin -> Multiplatform library 
There the setup for the runners is already done via kotlin-test and test-junit for JVM and nodejs for JS i.e. no need for mocha or jasmine nor Spek as far as I can tell. 
However, you need to add a testSetup.kt in JS in order that Atrium can pick up the correct implementation (for this you can copy testSetup.kt of atrium-fluent-en_GB-js)

File: samples/multiplatform/build.gradle [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            dependencies {
+                implementation kotlin('test-common')
+                implementation kotlin('test-annotations-common')
+                implementation ('ch.tutteli.atrium:atrium-fluent-en_GB-common:0.9.0-alpha2')
+            }
+        }
+        jvmTest {
+            dependencies {
+                //implementation kotlin('test')
+                //implementation kotlin('test-junit')
```
Comment (robstoll) [inline samples/multiplatform/build.gradle]: add those dependencies, needed for running tests

File: samples/multiplatform/build.gradle [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+                implementation("org.junit.jupiter:junit-jupiter-api:5.3.1")
+                implementation("ch.tutteli.atrium:atrium-fluent-en_GB:0.9.0-alpha2")
+            }
+        }
+        jsTest {
+            dependencies {
+                //implementation kotlin('test-js')
```
Comment (robstoll) [inline samples/multiplatform/build.gradle]: and add this dependency as well, needed for running tests

**Comment by robstoll** [comment]
Let me know in case you need more pointers

**Comment by bsemexan** [comment]
> @bsemexan Let me know in case you need more pointers

Yes, I did as suggested but seems the tests are not running

**Comment by robstoll** [comment]
Have you seen the latest reviews (uncomment some lines). This should already help a bit

**Comment by bsemexan** [comment]
> Have you seen the latest reviews (uncomment some lines). This should already help a bit

Yes, uncommented, and made the assertion changes so it would fail and nothing changed

**Comment by robstoll** [comment]
can you please push your changes, I'll take a look into it afterwards

**Comment by robstoll** [comment]
Thanks, I see two main problems:
- you use junit's `@Test` and not kotlin's add `import kotlin.test.Test`
- your structure is wrong, it should be src/jsTest/kotlin/... => I suggest you kick out one of them 
- same for jvm it should be src/jvmTest/kotlin => I suggest you kick out spek

**Comment by bsemexan** [comment]
> * import kotlin.test.Test

ok the JVM tests seem to be working properly, as they fail when the assertions are incorrect 
but the js tests are still passing

**Comment by robstoll** [comment]
right, you have to add nodejs to your js config

File: samples/multiplatform/build.gradle [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    mavenCentral()
+}
+group 'com.atrium'
+version '0.0.1'
+
+apply plugin: 'maven-publish'
+
+kotlin {
+    jvm()
+    js()
```
Comment (robstoll) [inline samples/multiplatform/build.gradle]:
```suggestion
    js{
      nodejs { }
    }
```

File: samples/multiplatform/src/jsTest/jasmine/kotlin/ch/tutteli/SmokeTest.kt [hunk 14]
```diff
@@ -0,0 +1,70 @@
+package ch.tutteli
+
+import ch.tutteli.atrium.api.fluent.en_GB-js.*
```
Comment (robstoll) [inline samples/multiplatform/src/jsTest/jasmine/kotlin/ch/tutteli/SmokeTest.kt]:
```suggestion
import ch.tutteli.atrium.api.fluent.en_GB.*
```

**Comment by robstoll** [comment]
Once you made these changes the tests will run and fail. You will see that you need to setup Atrium in addition. This is described in the js sample projects, for instance: https://github.com/robstoll/atrium/blob/master/samples/js/mocha/build.gradle#L81

**Comment by bsemexan** [comment]
> Once you made these changes the tests will run and fail. You will see that you need to setup Atrium in addition. This is described in the js sample projects, for instance: https://github.com/robstoll/atrium/blob/master/samples/js/mocha/build.gradle#L81
> 
> (and don't forget to add a test to commonTest/src/kotlin -- you can copy for instance the SmokeTest of atrium-fluent-en_GB-common)

seems adding 
`implementation("ch.tutteli.atrium:atrium-fluent-en_GB-js:0.9.0-alpha2")` to the jsTest dependincie does the setup.... but then all tests fail

**Comment by robstoll** [comment]
Fail with something like `no implementation for CoreFactory found` right? if that's the case then this exactly the missing setup I mean. Have a look at the link I posted. Easiest is to add the testSetup.kt (you can even copy an existing one, e.g. bundles/cc-en_GB-robstoll/atrium-cc-en_GB-robstoll-js/src/test/kotlin/testSetup.kt)

**Comment by bsemexan** [comment]
> Fail with something like `no implementation for CoreFactory found` right? if that's the case then this exactly the missing setup I mean. Have a look at the link I posted. Easiest is to add the testSetup.kt (you can even copy an existing one, e.g. bundles/cc-en_GB-robstoll/atrium-cc-en_GB-robstoll-js/src/test/kotlin/testSetup.kt)

```
* What went wrong:
Execution failed for task ':allTests'.
```

File: samples/multiplatform/build.gradle [hunk 15]
```diff
@@ -11,13 +11,16 @@ apply plugin: 'maven-publish'
...
+        nodejs {
+        }
+    }
     sourceSets {
         commonTest {
             dependencies {
                 implementation kotlin('test-common')
                 implementation kotlin('test-annotations-common')
-                implementation ('ch.tutteli.atrium:atrium-fluent-en_GB-common:0.9.0-alpha2')
+                implementation('ch.tutteli.atrium:atrium-fluent-en_GB-common:0.9.0-alpha2')
```
Comment (robstoll) [inline samples/multiplatform/build.gradle]:
```suggestion
                implementation('ch.tutteli.atrium:atrium-fluent-en_GB-common:0.9.1')
```

Comment (robstoll) [reply samples/multiplatform/build.gradle]: same same but different for other dependencies. Maybe it is best if you extract the versino number to a variable

**Comment by robstoll** [comment]
@bsemexan thanks for trying out 0.9.1, seems like this version is still not synced to maven central. I have to upload it again I guess. 0.9.0 probably fails as well due to https://github.com/robstoll/atrium/issues/338
Stay tuned, I don't have time any more today but I will upload a new version tomorrow or on the weekend

**Comment by robstoll** [comment]
bsemexan Finally the sync worked (took me 5 attempts, sometimes very annoying to sync artifacts from Bintray to Sonatype). Please retry it with Atrium 0.9.1

File: gradle/wrapper/gradle-wrapper.properties [hunk 16]
```diff
@@ -1,5 +1,6 @@
+#Tue Jan 21 11:00:06 EST 2020
```
Comment (robstoll) [inline gradle/wrapper/gradle-wrapper.properties]: please revert the changes in this file

File: samples/jvm/spek/gradlew.bat [hunk 17]
```diff
@@ -1,84 +1,84 @@
...
-rem Set variable GRADLE_EXIT_CONSOLE if you need the _script_ return code instead of
-rem the _cmd.exe /c_ return code!
-if  not "" == "%GRADLE_EXIT_CONSOLE%" exit 1
-exit /b 1
-
-:mainEnd
-if "%OS%"=="Windows_NT" endlocal
-
-:omega
+@if "%DEBUG%" == "" @echo off
```
Comment (robstoll) [inline samples/jvm/spek/gradlew.bat]: please revert the changes in this file and also other changes related to samples/jvm/spek, most likely that was a mistake

**Comment by robstoll** [comment]
@bsemexan the changes are not reverted yet. Shall I take over from here.?

**Comment by bsemexan** [comment]
> @bsemexan the changes are not reverted yet. You have changed the permissions of gradlew as it seems. Shall I take over from here?

yes

**Comment by robstoll** [comment]
Thanks for your contribution. How about another one, this time maybe more code related?
https://github.com/robstoll/atrium/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22
